atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStore.java

```diff
@@ -0,0 +1,151 @@
+     */
+    public synchronized Optional<Long> getUpperLimit() {
+        return cassandraKeyValueService.clientPool.runWithRetry(client -> {
+            ByteBuffer queryBuffer = CassandraTimestampUtils.constructSelectTimestampBoundQuery();
+            try {
+                CqlResult result = client.execute_cql3_query(queryBuffer, Compression.NONE, ConsistencyLevel.QUORUM);
+                List<CqlRow> cqlRows = result.getRows();
+                if (cqlRows.isEmpty()) {
+                    return Optional.<Long>empty();
+                }
+                return Optional.of(CassandraTimestampUtils.getLongValue(result));
+            } catch (TException e) {
+                throw Throwables.rewrapAndThrowUncheckedException(e);
+            }
+        });
+    }
+
+    /**
+     * Stores an upper timestamp limit in the database in an atomic way.
+     * @param expected Expected current value of the timestamp bound (or null if there is no bound)
+     * @param target Upper timestamp limit we want to store in the database
+     * @throws ConcurrentModificationException if the expected value does not match the bound in the database
+     */
+    public synchronized void storeTimestampBound(Long expected, long target) {
+        cassandraKeyValueService.clientPool.runWithRetry(client -> {
+            ByteBuffer queryBuffer = CassandraTimestampUtils.constructCheckAndSetCqlQuery(expected, target);
+            try {
+                CqlResult result = client.execute_cql3_query(queryBuffer, Compression.NONE, ConsistencyLevel.QUORUM);
+                checkOperationWasApplied(expected, target, result);
+                return null;
+            } catch (TException e) {
+                throw Throwables.rewrapAndThrowUncheckedException(e);
+            }
+        });
+    }
+
+    /**
+     * Writes a backup of the existing timestamp to the database. After this backup, this timestamp service can
+     * no longer be used until a restore.
+     *
+     * This may be implemented as follows:
+     *  - Read the value of the backup timestamp. If this is readable, skip (already backed up).
+     *  - Read the value of the timestamp (TS). If this is unreadable, fail.
+     *  - Do a conditional logged batch:
+     *    - CAS the main timestamp, expecting TS, to the INVALIDATED_VALUE
+     *    - Put unless exists the value of TS to the backup timestamp.
+     */
+    public synchronized void backupExistingTimestamp() {
+        throw new UnsupportedOperationException("TODO implement me");
+    }
+
+    /**
+     * Restores a backup of an existing timestamp.
+     *
+     * This may be implemented following the inverse of the backup process:
+     *  - Read the value of the timestamp. If this is readable, skip.
+     *  - Read the value of the backup (BT). If this is unreadable, fail.
+     *  - Do a conditional logged batch:
+     *    - Conditionally delete the backup if it matches BT
```

Comment:
Yes, it's a cleaner solution (otherwise I have to write a delete command in the timestamp bound store / timestamp utils classes, which has its own syntax). Will change to this.